### ISSUE TITLE
Time shift in FCAL showers?

### ISSUE THREAD

**Issue by sdobbs** [JeffersonLab/HDGeant4#10]
The mean reconstructed time for FCAL showers shows a ~1 ns shift.
[Hist_GenReconTrackComparison-FCALenergy.pdf](https://github.com/JeffersonLab/HDGeant4/files/926815/Hist_GenReconTrackComparison-FCALenergy.pdf)

Judging from the delta p vs. p distribution, it appears that there might be an energy dependence in this shift, which means that it could possible be due to some difference in the reported depths at which the shower occurs?

**Comment by mashephe** [comment]
There is an energy dependent correction to the time that is meant to compensate for depth in the code.

In the data we observe a systematic shift in time with energy.  I suspect this correction isn't properly tuned or there really is a systematic shift with energy, e.g., time walk in the FADC.  Adesh is examining this in detail.

Matt

**Comment by sdobbs** [comment]
Right, we certainly see a systematic shift in data, but I think it's worth looking into the source of the systematic shift of this size between two versions of hdgeant before simply calibrating it away.

**Comment by sdobbs** [comment]
Below is a distribution of the difference between the FCAL shower truth time and the truth hit times for the hits in a 3x3 region around the central block of the shower.

Red is HDGeant4, blue is HDGeant.

![histogram_fcaltdiff](https://user-images.githubusercontent.com/1182058/42784860-16e526b8-8916-11e8-8f06-9317953511ba.png)

**Comment by rjones30** [comment]
I traced this discrepancy down to a difference in the two the algorithms stitch together multiple hits from a shower to build an aggregate pulse in a block. The behavior in hdgeant (energy weighted average) is more realistic than the earliest-time algorithm that was used in hdgeant4. The fadc250 uses a constant-fraction technique for locating the leading edge of a pulse, which is equivalent to an energy-weighted average time. I changed the hdgeant4 algorithm to agree with hdgeant. To get this new behavior, a total of 3 new pull requests need to be incorporated, 2 for bugs I found along the way, and the latest one to fix this problem. Below is a new comparison between the hit-shower times for hdgeant (navy) and hdgeant4 (red). The remaining difference comes from slight differences in the shower propagation in the G3 and G4 libraries, and the lower energy cutoffs in G4 tracking.

![image](https://user-images.githubusercontent.com/7832920/43133496-f6f8a9e0-8f0c-11e8-91d2-d4dc7d204c00.png)
